### ISSUE TITLE
Can't set breakpoint in Android Studio

### ISSUE THREAD

**Issue by In-line** [intellij-rust/intellij-rust#3399]
## Environment

* **Intellij-Rust plugin version:** 0.2.0.2114-182
* **Rust toolchain version:**: rustc 1.33.0-beta.6 (b203178b6 2019-02-05)
* **IDE name and version:** Android Studio 3.3.1
* **Operating system:**: Mac/Linux

## Problem description

Can't set breakpoint in Android Studio in opened *.rs file. Simple clicking just didn't work. 

## Steps to reproduce

Install intellij-rust, setup rust project to use with Android, try to set breakpoint in any rust source.

**Comment by Undin** [comment]
It's expected because debugger is supported only in CLion at this moment

**Comment by In-line** [comment]
@Undin Where can I fix this "expected" issue? Is it a problem with Android Studio or rust-intellij?

Android Studio already supports debugging of NDK code, so for me personally this is bug and it should be fixed.

**Comment by Undin** [comment]
@In-line It's not a bug of Android Studio or Intellij-Rust. It's a feature that it's not supported at this moment.
We haven’t had plans to implement it yet. At this moment this feature has low priority because there aren't a lot of requests for it. However, if community interest on it grows, we will pay more attention to it.

**Comment by andreytkachenko** [comment]
@Undin how this interest was measured, is any topic to read related to that, as for me, I am not using intellij-rust (and forced to use vscode) only because of absence of debugging (CLion not count - it is not community).

**Comment by Undin** [comment]
> @Undin how this interest was measured

@andreytkachenko for example, by positive reactions and/or by comments in this issue

> CLion not count - it is not community

Note, this issue is not about debugger in community IDEs, only in AS

**Comment by vlad20012** [comment]
> how this interest was measured

@andreytkachenko We have less then 0.5% users in AS

> I am not using intellij-rust (and forced to use vscode) only because of absence of debugging (CLion not count - it is not community).

You can also take a look at [CLion EAP](https://www.jetbrains.com/clion/nextversion/). It's available for free.

> Where can I fix this "expected" issue? Is it a problem with Android Studio or rust-intellij?

@In-line Personally I know nothing about AS and its debugger, but all our debugger-related sources are located [here](https://github.com/intellij-rust/intellij-rust/tree/master/debugger).

Note that debugger is a **proprietary part** of AS/CLion, so even if enabling debugger support in AS is relatively easy (which I personally doubt), we can't just enable it without permission from our management... Or at least I'm not sure about that.

**Comment by In-line** [comment]
> @andreytkachenko for example, by positive reactions and/or by comments in this issue
So it is our mission to spread information about this issue to the Rust community.

**Comment by arshakmanukyan** [comment]
Rust is the good replacement for C++ NDK development. Not having ability to debug is the blocker issue for me. I rather use VSCode than intellij-rust.

**Comment by vlad20012** [comment]
Btw does VSCode now have a good debugger for rust?

**Comment by andreytkachenko** [comment]
@vlad20012 it is not good, but it at least works, pretty-printer may be much better, sometimes it doesn't showing local variables (not sure may be it is lldb problem), but anyway it can save time (especially when dealing with microcontrollers - there it works well)

**Comment by andreytkachenko** [comment]
@vlad20012 
> You can also take a look at CLion EAP. It's available for free.
I didn't mean that I need free software, I said that there is a request for a community debugger from users, located at github, bitbucket or wherever and accepting PRs. I do thing that it is not a rocket science to implement wrapper to lldb or gdb and make some interface in IDE (like it already made for pycharm for example).

**Comment by MatrixDev** [comment]
Are there at least some plans on resolving this issue? Rust is becoming more and more popular but some projects just can't start switching to it without the debugger support.

**Comment by Undin** [comment]
@MatrixDev there are no plans to support debugging in Android Studio currently. Debugging is available in CLion, IntelliJ IDEA Ultimate, PyCharm Pro and GoLand. See https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#compatible-ids for more details about plugin capabilities in different IDEs

**Comment by Didgy74** [comment]
I'd like to voice my desire for this to be implemented.

I'm currently making a native application for Android and wish to rewrite the project in Rust. But as long as I can't have a useful debugging experience, I have to put it on hold.

**Comment by MatrixDev** [comment]
This is such a bummer...

If I have a crash while the debugger is attached in the Android Studio - I'm automatically navigated to the actual file and all variables are watchable.
Settings breakpoints from LLDB command line works as well.

So debugger does work in Android Studio. Only thing that is blocking at least somewhat usable experience - it is allowing actual breakpoints setting in the UI. I don't thing there should be any "legal" reasons against it.

**Comment by minyung** [comment]
I set the bp in intellij. 
When I opened the project in Android Studio, bp was set, and debugging worked fine. 
Remove breakpoints in Android studio also work...

It seems that only breakpoints can't be set.

**Comment by MatrixDev** [comment]
Yes, it's just like mouse clicks are not handled in Rust for AS. This should be trivial to fix and has nothing to do with the debugger itself.

**Comment by hangedfish** [comment]
Is there any good news?

**Comment by xanahopper** [comment]
>  We have less then 0.5% users in AS

Because users recognize intellij-rust in AS is almost unusable, IT CAN'T DEBUG. Then we were told it should have enough users to use it. It is a dead-lock, like chicken and egg.

**Comment by MatrixDev** [comment]
> > We have less then 0.5% users in AS
> 
> Because users recognize intellij-rust in AS is almost unusable, IT CAN'T DEBUG. Then we were told it should have enough users to use it. It is a dead-lock, like chicken and egg.

Agree. It doesn't have a basic functionality - neither debugging (IMHO Rust plugin issue) nor dedicated source sets (Android plugin). I wonder, why user numbers are so low?...

**Comment by xanahopper** [comment]
The funny thing is that after some tring, with the latest stable AS and Intellij-Rust, we can launch a dual debug in AS and set breakpoints in Rust.
- Cargo/Rust build target should be included in any way: `cargo-android-plugin`, `gradle-cargo-plugin`, or despite build and cp manually.
- Avoid strip cargo build target(the debugger needs the symbols) by adding `build.gradle` config like `doNotStrip`.
- Find the rust source code/project, and attach it.
- Run a dual debug, set breakpoints in Rust.

It works.

**Comment by kl** [comment]
I made a plugin that enables you to set line breakpoints in Rust files in Android Studio by clicking the gutter or using the toggle breakpoint keyboard shortcut: https://github.com/kl/gutter-breakpoints-rust-plugin
